app/posthandler/fee_refund_decorator.go

```diff
@@ -0,0 +1,127 @@
+package posthandler
+
+import (
+	"fmt"
+
+	"cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	authkeeper "github.com/cosmos/cosmos-sdk/x/auth/keeper"
+	"github.com/cosmos/cosmos-sdk/x/auth/types"
+	feegrantkeeper "github.com/cosmos/cosmos-sdk/x/feegrant/keeper"
+)
+
+const (
+	eventType                   = "fee_refund"
+	attributeKeyRefund          = "refund"
+	attributeKeyRefundRecipient = "refund_recipient"
+	bondDenom                   = "utia"
+)
+
+// FeeRefundDecorator handles refunding a portion of the fee that was originally
+// deducted from the feepayer but was not needed because the tx consumed less
+// gas than the gas limit.
+type FeeRefundDecorator struct {
+	accountKeeper  authkeeper.AccountKeeper
+	bankKeeper     types.BankKeeper
+	feegrantKeeper feegrantkeeper.Keeper
+}
+
+func NewFeeRefundDecorator(ak authkeeper.AccountKeeper, bk types.BankKeeper, fk feegrantkeeper.Keeper) FeeRefundDecorator {
+	return FeeRefundDecorator{
+		accountKeeper:  ak,
+		bankKeeper:     bk,
+		feegrantKeeper: fk,
+	}
+}
+
+func (frd FeeRefundDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	if err := frd.maybeRefund(ctx, tx); err != nil {
+		return ctx, err
+	}
+	return next(ctx, tx, simulate)
+}
+
+func (frd FeeRefundDecorator) maybeRefund(ctx sdk.Context, tx sdk.Tx) error {
+	// Replace the context's gas meter with an infinite gas meter so that this
+	// decorator doesn't run out of gas while refunding.
+	gasMeter := ctx.GasMeter()
+	ctx = ctx.WithGasMeter(sdk.NewInfiniteGasMeter())
+
+	feeTx, ok := tx.(sdk.FeeTx)
+	if !ok {
+		return errors.Wrap(sdkerrors.ErrTxDecode, "tx must be a FeeTx to use FeeRefundDecorator")
+	}
+
+	if gasMeter.IsOutOfGas() {
+		// If the gas meter is out of gas, then no refund needs to be issued.
+		return nil
+	}
+
+	coinsToRefund := getCoinsToRefund(gasMeter, feeTx)
+	refundRecipient := getRefundRecipient(feeTx)
+	refundRecipientAccount := frd.accountKeeper.GetAccount(ctx, refundRecipient)
+	if refundRecipientAccount == nil {
+		return errors.Wrapf(sdkerrors.ErrUnknownAddress, "refund recipient address: %s does not exist", refundRecipientAccount)
+	}
+
+	if err := frd.processRefund(frd.bankKeeper, ctx, refundRecipientAccount, coinsToRefund); err != nil {
+		return err
+	}
+
+	events := sdk.Events{newFeeRefundEvent(coinsToRefund, refundRecipient)}
```

Comment:
I'm not 100% sure adding an extra event to each tx is worth it, especially since all the information that a user/wallet needs is already contained in gas used

Reply:
26fc642efa31b9e176a30c6652ff0dbcb63779b5